rfcs/134-write-behavior-consistency.md

```diff
@@ -0,0 +1,234 @@
+
+For `Copy` and `Move`:
+
+`ipfs` have native support for `Copy` and `Move`. If the dst exists, the `Copy` operation will return `FileExists`, while the `Move` operation will overwrite.
+
+### dropbox
+
+For `Write`:
+
+`dropbox` supports setting `WriteMode`. With `files.WriteModeAdd`, `Write` will not return `FileExists` too:
+
+```go
+input := &files.CommitInfo{
+    Path: rp,
+    Mode: &files.WriteMode{
+        Tagged: dropbox.Tagged{
+            Tag: files.WriteModeAdd,
+        },
+    },
+}
+```
+
+For `Append`:
+
+`dropbox` supports `WriteMode` like `Write`.
+
+For `Copy` and `Move`:
+
+`dropbox` supports `autorename`.
+
+- If `autorename` is false, `dropbox` will return `to/conflict/file/` error.
+- If `autorename` is true, `dropbox` will rename the final path with a suffix like `(1)`.
+
+
+In [GSP-46](https://github.com/beyondstorage/specs/blob/master/rfcs/46-idempotent-delete.md), we make `Delete` idempotent which means without any outsides changes, any `Delete` call on the same path will get the same result. But `Write` is more complex than `Delete` and we cannot reuse the experience from GSP-46 directly:
+
+- `Write` is unlikely to be retried: the underlying `io.Reader` could only be consumed once.
+- Make `Delete` idempotent is simple: we just need to ignore the `ObjectNotExist` error, but it's difficult for `Write`.
+
+## Proposal
+
+I propose to improve write behavior consistency in the following method:
+
+- All write operations SHOULD NOT return an error as the object exists.
+- All successful write operations SHOULD be complete
+
+Compared to GSP-46, we will not enforce the requirement for idempotent, because most write operations are not idempotent. For example, the second call for `Move(a, b)` should report `ObjectNotExist`.
+
+`Complete` means after a successful write, the object's content and metadata should be the same as specified in write request. For example, write `1024` bytes on an existing `2048` bytes file should result in a `1024` bytes file.
+
+### Write Operations
+
+Write operations are those that will create objects in the service, including:
+
+- `Write`
+- `Copy`
+- `Move`
+- `Fetch`
+- `CreateDir`
+- `CreateLink`
+- `CreateMultipart`
+- `CompleteMultipart`
+- `CreateBlock`
+- `CombineBlock`
+- `CreateAppend`
+- `CommitAppend`
+- `CreatePage`
+
+Other APIs are out of scope.
+
+### Services Implementations
+
+- Service that has native support for `overwrite` or `create` doesn't NEED to check the object exists or not.
+- Service that doesn't have native support for `overwrite` or `create` SHOULD check and delete the object if exists.
+  - `autoreanme` alike features are forbidden: service provider SHOULD NOT change the final destination path.
+- Service SHOULD make sure write operations complete.
+
+### Error Handling
+
+If user call `Write(src, dst)` but dst is a dir, service SHOULD return `ErrObjectModeInvalid` error instead of
+
+- removing the dst
+- returning `ErrObjectExist`
+- write into `dst/src`
```

Comment:
Do we need `Stat(dst)` inside `Write(src, dst)` to get whether `dst` is an existing dir?

Reply:
No, we don't need to.

If `dst` is an existing dir, we should return `ErrObjectModeInvaild`.

Reply:
Got it.